File: 0187_repeated_dna_sequences/hash_set/python/solution.py
WINDOW_SIZE = 10


def find_repeated_dna_sequences(s: str) -> list[str]:
    if len(s) < WINDOW_SIZE:
        return []

    seen: set[str] = set()
    repeated: set[str] = set()
    result: list[str] = []

    for start in range(len(s) - WINDOW_SIZE + 1):
        fragment = s[start:start + WINDOW_SIZE]
        if fragment in seen:
            if fragment not in repeated:
                repeated.add(fragment)
                result.append(fragment)
        else:
            seen.add(fragment)

    return result

File: 0187_repeated_dna_sequences/hash_set/kotlin/solution.kt
private const val WINDOW_SIZE = 10

fun findRepeatedDnaSequences(s: String): List<String> {
    if (s.length < WINDOW_SIZE) return emptyList()

    val seen = mutableSetOf<String>()
    val repeated = mutableSetOf<String>()
    val result = mutableListOf<String>()

    for (start in 0..s.length - WINDOW_SIZE) {
        val fragment = s.substring(start, start + WINDOW_SIZE)
        if (fragment in seen) {
            if (repeated.add(fragment)) {
                result.add(fragment)
            }
        } else {
            seen.add(fragment)
        }
    }

    return result
}
